### DISCUSSION TITLE
Does `AsChangeset` skip referenced fields and primary key?

### DISCUSSION THREAD

**Discussion by commonuserlol** [diesel-rs/diesel#4355] [Q&A] [answered]
My SQL is:
```sql
CREATE TABLE foo (
    id INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT,
    thing INTEGER
);

CREATE TABLE foo_child (
    id INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT,
    something INTEGER,
    foo_id INTEGER NOT NULL REFERENCES foo(id)
)
```
and corresponding structs:
```rust
#[derive(Queryable, Selectable, Insertable, Identifiable, PartialEq, Debug, AsChangeset)]
#[diesel(table_name = crate::database::schema::foo)]
#[diesel(check_for_backend(diesel::sqlite::Sqlite))]
pub struct Foo {
    pub id: i32,
    pub thing: Option<i32>
}

#[derive(Queryable, Selectable, Insertable, Identifiable, Associations, Debug, PartialEq, Default)]
#[diesel(belongs_to(Foo))]
#[diesel(table_name = crate::database::schema::foo_child)]
#[diesel(check_for_backend(diesel::sqlite::Sqlite))]
pub struct Child {
    pub id: i32,
    pub something: Option<i32>,
    pub foo_id: i32,
}
```

so my question is, when I create new row of `foo_child` using `insert_into`, I use `Default` for every field except `foo_id`. Does it mean `id` will be overriden to `0` (default) or diesel/sqlite will set correct value? Same question applies to `update`.
```rust
insert_into(foo_child::table)
            .values(&Child {
                foo_id: value.id,
                ..Default::default()
            })
            .execute(connection)
            .map_err(|e| e.to_string())?;
```
Or there's better way to do it?

Sorry if it's a duplicate.

**Comment by commonuserlol** [comment]
Well, no. Diesel doesn't override `id` value. Can I `#[derive(Identifiable)]` without `id` field in schema?

**Comment by weiznich** [comment, marked as answer]
Diesel generally inserts/updates all fields to the provided values. So if you set something to 0 diesel will set the database field to zero. The only exception there are primary key fields for updates. 
In addition to that both Insertsable and AsChangeset have special behaviour for optional fields. If such a field is None for an insert statement the database side default value is used. For updates the field is skipped. 

The Identifiable derive is usefully for identifying a database entry, which requires a primary key value. For that by default the id field is used but you can specify one or more other fields via the `#[diesel(primary_key())]` attribute. 

All of this is explained in depth in the [All About Inserts](http://diesel.rs/guides/all-about-inserts.html) and [All about Updates](http://diesel.rs/guides/all-about-updates.html) guides and the API documentation of the relevant derives.

**Reply by commonuserlol** [reply]
So I can use `id INTEGER PRIMARY KEY AUTOINCREMENT` (aka `Option<i32>`) and leave it as `None` for autoincrementing?

**Reply by weiznich** [reply]
Yes

**Reply by commonuserlol** [reply]
Awesome, thanks a lot!

**Reply by commonuserlol** [reply]
Sorry for bothering you again, but when primary key is nullable, `belonging_to` is not working. Perhaps there's other way to keep primary key `NOT NULL` and still increment key while using struct as changeset?

**Reply by weiznich** [reply]
Well, now you mix in even another derive. They can work together but the more derives you add the more requirements on the struct layout you get. For `#[derive(Associations)]` a compatible primary key type is required. A nullable value is not compatible to a not nullable database value.

The preferred way to handle that is to use separate structs for separate usecases. That helps you to avoid situations where you suddenly need to unwrap a primary key field or something like that.
